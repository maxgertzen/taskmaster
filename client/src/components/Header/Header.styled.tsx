import styled from '@emotion/styled';

export const HeaderContainer = styled.header(({ theme }) => ({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  padding: theme.spacing(2),
  backgroundColor: theme.colors.surface,
  [`@media (max-width: ${theme.breakpoints.sm})`]: {
    padding: theme.spacing(1),
    height: '10vh',
  },
}));

export const UserActionsContainer = styled.div(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  span: {
    marginRight: theme.spacing(1),
  },
  [`@media (max-width: ${theme.breakpoints.sm})`]: {
    flexDirection: 'row-reverse',
    justifyContent: 'space-between',
    alignItems: 'center',
    position: 'fixed',
    top: 0,
    left: 0,
    width: '100vw',
    backgroundColor: theme.colors.surface,
    zIndex: 1000,
    gap: theme.spacing(1),
    padding: theme.spacing(2),
    height: 'inherit',
    span: {
      marginRight: 0,
      flex: 1,
    },
  },
}));

export const HamburgerMenu = styled.div(({ theme }) => ({
  display: 'none',
  [`@media (max-width: ${theme.breakpoints.sm})`]: {
    display: 'block',
    cursor: 'pointer',
    padding: theme.spacing(1),
    zIndex: 1100,
  },
}));

export const UserMenuContainer = styled.div(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  marginLeft: theme.spacing(1),
  [`@media (max-width: ${theme.breakpoints.sm})`]: {
    order: -1,
    gap: theme.spacing(2),
  },
}));

export const VerticalDivider = styled.div(({ theme }) => ({
  width: '1px',
  height: '24px',
  backgroundColor: theme.colors.text,
  margin: theme.spacing(0, 2),
  [`@media (max-width: ${theme.breakpoints.sm})`]: {
    display: 'none',
  },
}));

export const UserNameContainer = styled.span(({ theme }) => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  position: 'relative',
  [`@media (max-width: ${theme.breakpoints.sm})`]: {
    flexDirection: 'row-reverse',
    gap: theme.spacing(1),
  },
}));

export const BackButtonWrapper = styled.div({
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  flex: 1,
});

export const UserActionPanelViewContainer = styled.div(({ theme }) => ({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'center',
  width: '100%',
  height: 'inherit',
  padding: theme.spacing(1),
}));
