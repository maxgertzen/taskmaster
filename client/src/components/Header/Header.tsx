import { useThemeStore } from '../../store/themeStore';
import { User } from '../../types/shared';
import { FaIcon } from '../FontAwesomeIcon/FontAwesomeIcon';
import { Logo } from '../Logo/Logo';

import {
  HeaderContainer,
  UserActionsContainer,
  DarkModeIcon,
  VerticalDivider,
} from './Header.styled';

interface HeaderProps {
  user: User;
}

export const Header: React.FC<HeaderProps> = ({ user }) => {
  const activeTheme = useThemeStore((state) => state.theme);
  const toggleTheme = useThemeStore((state) => state.toggleTheme);

  return (
    <HeaderContainer>
      <Logo withTitle />
      <UserActionsContainer>
        <DarkModeIcon
          icon={['far', activeTheme === 'light' ? 'moon' : 'sun']}
          size='lg'
          onClick={toggleTheme}
        />
        <VerticalDivider />
        <span>
          Hi, <b>{user.name}</b>
        </span>
        <FaIcon
          icon={[activeTheme === 'light' ? 'fas' : 'far', 'user-circle']}
          size='lg'
        />
      </UserActionsContainer>
    </HeaderContainer>
  );
};
