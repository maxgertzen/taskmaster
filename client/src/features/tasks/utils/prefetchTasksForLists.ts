import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
