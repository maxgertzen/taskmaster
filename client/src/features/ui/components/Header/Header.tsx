import { Fragment, useRef } from 'react';

import { usePopupMenuState } from '@/shared/hooks';
import { useDashboardStore } from '@/shared/store/dashboardStore';
import { useViewportStore } from '@/shared/store/viewportStore';
import { UserDetails } from '@/shared/types/shared';
import { SelectedView } from '@/shared/types/ui';

import {
  Logo,
  LogoutButton,
  PopupMenu,
  SpriteIcon,
  SwitchThemeButton,
  Searchbar,
} from '..';

import {
  BackButtonWrapper,
  HeaderContainer,
  UserActionPanelViewContainer,
  UserActionsContainer,
  UserNameContainer,
  VerticalDivider,
} from './Header.styled';

interface HeaderProps {
  user: UserDetails | null;
  view?: SelectedView;
  setView?: (view: SelectedView) => void;
  onBack?: () => void | false;
}

const UserName: React.FC<HeaderProps> = ({ user }) => {
  return (
    <span>
      Hi
      {user?.name ? (
        <>
          , <b>{user.name}!</b>
        </>
      ) : (
        '!'
      )}
    </span>
  );
};

export const Header: React.FC<HeaderProps> = ({
  user,
  view,
  setView,
  onBack,
}) => {
  const triggerRef = useRef<HTMLDivElement>(null);
  const { closeMenu, isOpen, toggleMenu } = usePopupMenuState();
  const selectedList = useDashboardStore((state) => state.selectedList);
  const setSelectedList = useDashboardStore((state) => state.setSelectedList);

  const isMobile = useViewportStore((state) => state.isMobile);

  return (
    <HeaderContainer>
      {!isMobile && <Logo withTitle />}
      {onBack && isMobile ? (
        <UserActionPanelViewContainer>
          <BackButtonWrapper
            role='button'
            aria-label='Back to lists'
            onClick={onBack}
          >
            <SpriteIcon name='arrowButton' size={4} />
            <label>Lists</label>
          </BackButtonWrapper>
          <Searchbar
            placeholder='Search tasks'
            selectedListId={selectedList?.id}
            onSearchCallback={() => {
              setSelectedList(null);
              setView?.('board');
            }}
          />
        </UserActionPanelViewContainer>
      ) : (
        <UserActionsContainer>
          <Searchbar
            placeholder='Search tasks'
            selectedListId={selectedList?.id}
            onSearchCallback={() => setSelectedList(null)}
          />
          {!isMobile && (
            <Fragment>
              <SwitchThemeButton />
              <VerticalDivider />
            </Fragment>
          )}

          <UserNameContainer>
            {view === 'board' && isMobile ? (
              <BackButtonWrapper
                role='button'
                aria-label='Back to lists'
                onClick={onBack}
              >
                <SpriteIcon name='arrowButton' size={4} />
                <label>Lists</label>
              </BackButtonWrapper>
            ) : (
              <Fragment>
                <UserName user={user} />
                <div
                  ref={triggerRef}
                  aria-haspopup='true'
                  aria-expanded={isOpen}
                  onClick={toggleMenu}
                >
                  <PopupMenu
                    options={[
                      { label: <SwitchThemeButton withText /> },
                      { label: <LogoutButton /> },
                    ]}
                    onClose={closeMenu}
                    isOpen={isOpen}
                    triggerRef={triggerRef}
                  />
                  <SpriteIcon name='user' size={4} />
                </div>
              </Fragment>
            )}
          </UserNameContainer>
        </UserActionsContainer>
      )}
    </HeaderContainer>
  );
};
