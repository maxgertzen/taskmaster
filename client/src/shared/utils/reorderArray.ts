export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
