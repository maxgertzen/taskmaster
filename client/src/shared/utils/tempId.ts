export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
