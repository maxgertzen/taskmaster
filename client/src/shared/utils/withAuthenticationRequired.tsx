import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
