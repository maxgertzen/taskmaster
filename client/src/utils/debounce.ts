export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
