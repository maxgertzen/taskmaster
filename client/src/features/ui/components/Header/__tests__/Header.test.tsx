import { fireEvent } from '@testing-library/dom';
import { act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';

import {
  mockDashboardStore,
  mockViewportStore,
  mockPopupMenuState,
  mockIsMobile,
  mockSetSelectedList,
  mockSetSearchTerm,
} from '@/tests/__mocks__/storeMocks';
import { render, screen } from '@/tests/utils';

import { Header } from '../Header';

vi.mock('@/shared/store/dashboardStore', () => ({
  useDashboardStore: mockDashboardStore,
}));

vi.mock('@/shared/store/viewportStore', () => ({
  useViewportStore: mockViewportStore,
}));

vi.mock('@/shared/hooks', () => ({
  usePopupMenuState: mockPopupMenuState,
}));

describe('Header', () => {
  const mockUser = {
    id: '1',
    name: 'John Doe',
    email: 'john@example.com',
  };

  const mockSetView = vi.fn();
  const mockOnBack = vi.fn();

  beforeEach(async () => {
    vi.clearAllMocks();
  });

  // Desktop View Tests
  describe('Desktop View', () => {
    beforeEach(() => {
      mockIsMobile.mockReturnValue(false);
    });

    it('renders correctly in light theme', () => {
      const { container } = render(
        <Header
          user={mockUser}
          view='board'
          setView={mockSetView}
          onBack={mockOnBack}
        />,
        { theme: 'light' }
      );
      expect(container).toMatchSnapshot();
    });

    it('renders correctly in dark theme', () => {
      const { container } = render(
        <Header
          user={mockUser}
          view='board'
          setView={mockSetView}
          onBack={mockOnBack}
        />,
        { theme: 'dark' }
      );
      expect(container).toMatchSnapshot();
    });

    it('displays logo with title', () => {
      render(<Header user={mockUser} />);
      expect(screen.getByTestId('logo')).toBeInTheDocument();
    });
  });

  // Mobile View Tests
  describe('Mobile View', () => {
    beforeEach(() => {
      mockIsMobile.mockReturnValue(true);
    });

    it('renders correctly in light theme', () => {
      const { container } = render(
        <Header
          user={mockUser}
          view='board'
          setView={mockSetView}
          onBack={mockOnBack}
        />,
        { theme: 'light' }
      );
      expect(container).toMatchSnapshot();
    });

    it('renders correctly in dark theme', () => {
      const { container } = render(
        <Header
          user={mockUser}
          view='board'
          setView={mockSetView}
          onBack={mockOnBack}
        />,
        { theme: 'dark' }
      );
      expect(container).toMatchSnapshot();
    });

    it('does not display logo', () => {
      render(<Header user={mockUser} />);
      expect(screen.queryByTestId('logo')).toBeNull();
    });
  });

  // User Details Tests
  describe('User Details', () => {
    it('displays user name when provided', () => {
      render(<Header user={mockUser} />);
      expect(screen.getByTestId('user-name')).toBeInTheDocument();
    });

    it('displays generic greeting when no user name is provided', () => {
      render(<Header user={{ ...mockUser, name: '' }} />);
      expect(screen.getByTestId('user-name')).toBeInTheDocument();
    });

    it('handles null user', () => {
      render(<Header user={null} />);
      expect(screen.queryByTestId('user-name')).not.toBeNull();
    });
  });

  // View Tests
  describe('View Behavior', () => {
    beforeEach(() => {
      mockIsMobile.mockReturnValue(true);
    });

    it('handles board view with mobile back button', async () => {
      render(<Header user={mockUser} view='board' onBack={mockOnBack} />);

      const backButton = screen.getByTestId('back-to-lists');
      await act(async () => {
        fireEvent.click(backButton);
      });
      expect(mockOnBack).toHaveBeenCalled();
    });

    it('handles list view', () => {
      render(<Header user={mockUser} view='list' setView={mockSetView} />);
      expect(screen.queryByTestId('back-to-lists')).toBeNull();
    });
  });

  // Search Behavior Tests
  describe('Search Behavior', () => {
    it('handles search in board view with view change', async () => {
      vi.useFakeTimers();
      mockIsMobile.mockReturnValue(true);

      render(<Header user={mockUser} view='board' setView={mockSetView} />);
      const searchIcon = screen.getByTestId('search-icon');
      await act(async () => {
        fireEvent.click(searchIcon);
      });

      const input = screen.getByTestId('input');

      await act(async () => {
        fireEvent.change(input, { target: { value: 'test search' } });
      });

      await act(async () => {
        vi.runAllTimers();
      });

      expect(mockSetSearchTerm).toHaveBeenCalledWith('test search');
      expect(mockSetSelectedList).toHaveBeenCalledWith(null);

      vi.useRealTimers();
    });

    it('handles search without view change', async () => {
      vi.useFakeTimers();
      render(<Header user={mockUser} />);

      const searchIcon = screen.getByTestId('search-icon');

      await act(async () => {
        fireEvent.click(searchIcon);
      });

      const input = screen.getByTestId('input');

      await act(async () => {
        fireEvent.change(input, { target: { value: 'test search' } });
      });

      await act(async () => {
        vi.runAllTimers();
      });

      expect(mockSetSelectedList).toHaveBeenCalledWith(null);
      expect(mockSetView).not.toHaveBeenCalled();
    });
  });
});
