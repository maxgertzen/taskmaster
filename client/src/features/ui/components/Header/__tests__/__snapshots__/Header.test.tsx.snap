// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > Desktop View > renders correctly in dark theme 1`] = `
<div>
  <header
    class="css-1s39ib8"
    data-testid="header"
  >
    <div
      class="css-q3eegb"
    >
      <img
        alt="TaskMaster Logo"
        class="css-fhjm1v"
        data-testid="logo"
        src="/src/shared/assets/images/taskmaster.png"
      />
      <h1
        class="css-dytuun"
      >
        TaskMaster
      </h1>
    </div>
    <div
      class="css-f0nxor"
    >
      <div
        class="css-1egpccx"
        data-testid="searchbar"
      >
        <input
          class="css-1q2qwsy"
          data-testid="input"
          placeholder="Search tasks"
          type="text"
          value=""
        />
        <div
          class="css-1rqeb1j"
          data-spotlight=""
        >
          <div
            class="css-y8t6if"
            data-testid="search-icon"
            role="button"
            size="4"
          >
            <svg
              data-testid="test-svg"
            />
          </div>
        </div>
      </div>
      <div
        class="css-4dcg29"
        data-testid="icon-moon"
        role="button"
        size="3"
      >
        <svg
          data-testid="test-svg"
        />
      </div>
      <div
        class="css-1n1a5w1"
      />
      <span
        class="css-ijuw38"
      >
        <span
          data-testid="user-name"
        >
          Hi
          , 
          <b>
            John Doe
            !
          </b>
        </span>
        <div
          aria-expanded="false"
          aria-haspopup="true"
        >
          <div
            class="css-y8t6if"
            data-testid="icon-user"
            role="button"
            size="4"
          >
            <svg
              data-testid="test-svg"
            />
          </div>
        </div>
      </span>
    </div>
  </header>
</div>
`;

exports[`Header > Desktop View > renders correctly in light theme 1`] = `
<div>
  <header
    class="css-qbgrf2"
    data-testid="header"
  >
    <div
      class="css-v7ss51"
    >
      <img
        alt="TaskMaster Logo"
        class="css-fhjm1v"
        data-testid="logo"
        src="/src/shared/assets/images/taskmaster.png"
      />
      <h1
        class="css-dytuun"
      >
        TaskMaster
      </h1>
    </div>
    <div
      class="css-1jhy33"
    >
      <div
        class="css-1egpccx"
        data-testid="searchbar"
      >
        <input
          class="css-10md7mk"
          data-testid="input"
          placeholder="Search tasks"
          type="text"
          value=""
        />
        <div
          class="css-1rqeb1j"
          data-spotlight=""
        >
          <div
            class="css-y8t6if"
            data-testid="search-icon"
            role="button"
            size="4"
          >
            <svg
              data-testid="test-svg"
            />
          </div>
        </div>
      </div>
      <div
        class="css-4dcg29"
        data-testid="icon-moon"
        role="button"
        size="3"
      >
        <svg
          data-testid="test-svg"
        />
      </div>
      <div
        class="css-a14o6c"
      />
      <span
        class="css-ijuw38"
      >
        <span
          data-testid="user-name"
        >
          Hi
          , 
          <b>
            John Doe
            !
          </b>
        </span>
        <div
          aria-expanded="false"
          aria-haspopup="true"
        >
          <div
            class="css-y8t6if"
            data-testid="icon-user"
            role="button"
            size="4"
          >
            <svg
              data-testid="test-svg"
            />
          </div>
        </div>
      </span>
    </div>
  </header>
</div>
`;

exports[`Header > Mobile View > renders correctly in dark theme 1`] = `
<div>
  <header
    class="css-1s39ib8"
    data-testid="header"
  >
    <div
      class="css-1nfnfq9"
    >
      <div
        aria-label="Back to lists"
        class="css-10ei59g"
        data-testid="back-to-lists"
        role="button"
      >
        <div
          class="css-y8t6if"
          data-testid="icon-arrowButton"
          role="button"
          size="4"
        >
          <svg
            data-testid="test-svg"
          />
        </div>
        <label>
          Lists
        </label>
      </div>
      <div
        class="css-1egpccx"
        data-testid="searchbar"
      >
        <div
          class="css-1rqeb1j"
          data-spotlight=""
        >
          <div
            class="css-y8t6if"
            data-testid="search-icon"
            role="button"
            size="4"
          >
            <svg
              data-testid="test-svg"
            />
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header > Mobile View > renders correctly in light theme 1`] = `
<div>
  <header
    class="css-qbgrf2"
    data-testid="header"
  >
    <div
      class="css-1nfnfq9"
    >
      <div
        aria-label="Back to lists"
        class="css-10ei59g"
        data-testid="back-to-lists"
        role="button"
      >
        <div
          class="css-y8t6if"
          data-testid="icon-arrowButton"
          role="button"
          size="4"
        >
          <svg
            data-testid="test-svg"
          />
        </div>
        <label>
          Lists
        </label>
      </div>
      <div
        class="css-1egpccx"
        data-testid="searchbar"
      >
        <div
          class="css-1rqeb1j"
          data-spotlight=""
        >
          <div
            class="css-y8t6if"
            data-testid="search-icon"
            role="button"
            size="4"
          >
            <svg
              data-testid="test-svg"
            />
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;
